### PR TITLE
[SD] CUDA Graphs update

### PR DESCRIPTION
**Pull request by mwawrzos**
# What does this PR do ?

Restore CUDA Graphs support for SD.

**Collection**: multimodal

# Changelog 
- Update in the CUDA Graph callback to align changes in PL API.
- `MegatronLatentDiffusion.training_step` accepts `batch` input instead of `dataloader_iter` so CUDA Graphs are supported.
- Remove circular dependencies in `nemo/core/optim/megatron_fused_adam.py`.
- Add missing resblock_gn_groups parameter.
- Exception is raised when gradient reduction is misconfigured.
- Add synthetic data-loader for SD.
- Tensors are allocated directly on GPU to avoid synchronizations with host in few places.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

File: nemo/collections/nlp/modules/common/megatron/module.py [hunk 1]
```diff
@@ -38,10 +38,6 @@
 _BF16_TYPES = (torch.BFloat16Tensor, torch.cuda.BFloat16Tensor)
 
 
-def param_is_not_shared(param):
```
Comment (Victor49152) [inline nemo/collections/nlp/modules/common/megatron/module.py:41]: There is still usage of this function in clip_grads.py, for example Line # 97. I think this function basically goes to use_fsdp path. Please keep it or do global search for all usages

File: nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py [hunk 2]
```diff
@@ -1812,6 +1825,8 @@ def training_step(self, dataloader_iter):
             # async grad allreduce is not currently implemented for O1/autocasting mixed precision training
             # so we all-reduce gradients after the pipeline
             self.allreduce_gradients()  # @sangkug we think this is causing memory to blow up (hurts perf)
+        else:
+            raise ValueError("Either distributed_fused_adam or megatron_amp_O2 needs to be set if ddp_overlap is set")
```
Comment (Victor49152) [inline nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py:1829]: Is this guard necessary? It make default config of SD train fail, which worked fine on convergence before

**Review by Victor49152 — CHANGES_REQUESTED** [review]

File: nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py (hunk 2 above)
Comment (mwawrzos) [reply nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py:1829]: Without the guard, when 'ddp_overlap' is set ([which is true by default](https://github.com/NVIDIA/NeMo/blob/d656f226ea51b6cf962607ddfa9b9b2bdf44c5b2/nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py#L1738)) and an optimizer different than `distributed_fused_adam` or `megatron_amp_O2` is in use (e.g. `megatron_fused_adam`), the training loop saliently skips optimizer step.

I spent lots of time debugging it, so I added the guard.

Comment (Victor49152) [reply nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py:1829]: So we should turn the default ddp_overlap to False? I can change the config file then

File: nemo/collections/nlp/modules/common/megatron/clip_grads.py [hunk 3]
```diff
@@ -164,29 +160,24 @@ def clip_grad_norm_fp32(parameters, max_norm, norm_type=2, use_fsdp=False):
...
         if use_fsdp:
-            total_sharded_norm_cuda = torch.cuda.FloatTensor([float(total_sharded_norm)])
             # Sum norm of grad shards across data-parallel GPUs.
             torch.distributed.all_reduce(
-                total_sharded_norm_cuda,
-                op=torch.distributed.ReduceOp.SUM,
-                group=parallel_state.get_data_parallel_group(),
+                total_sharded_norm, op=torch.distributed.ReduceOp.SUM, group=parallel_state.get_data_parallel_group(),
             )
-            total_norm_cuda += total_sharded_norm_cuda
```
Comment (Victor49152) [inline nemo/collections/nlp/modules/common/megatron/clip_grads.py:176]: I might be missing something here. I understand that we can get rid of creating cuda tensor, but total_norm and total sharded norm were added together after all_reduce. This total_sharded_norm seems like ignored when we delete it?
And this is the error message when I try to run SDXL train with fsdp:
```
Error executing job with overrides: ['trainer.max_epochs=-1', 'trainer.max_steps=100', 'model.data.train.dataset_path=["/datasets/coyo/test.pkl"]']
Traceback (most recent call last):
  File "/opt/NeMo/examples/multimodal/text_to_image/stable_diffusion/sd_xl_train.py", line 98, in main
    trainer.fit(model)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 544, in fit
    call._call_and_handle_interrupt(
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 43, in _call_and_handle_interrupt
    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/launchers/subprocess_script.py", line 105, in launch
    return function(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 580, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 987, in _run
    results = self._run_stage()
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 1033, in _run_stage
    self.fit_loop.run()
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/fit_loop.py", line 205, in run
    self.advance()
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/fit_loop.py", line 363, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/training_epoch_loop.py", line 140, in run
    self.advance(data_fetcher)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/training_epoch_loop.py", line 250, in advance
    batch_output = self.automatic_optimization.run(trainer.optimizers[0], batch_idx, kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 190, in run
    self._optimizer_step(batch_idx, closure)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 268, in _optimizer_step
    call._call_lightning_module_hook(
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 157, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
  File "/opt/NeMo/nemo/collections/nlp/models/language_modeling/megatron_base_model.py", line 1260, in optimizer_step
    super().optimizer_step(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py", line 1303, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/optimizer.py", line 152, in step
    step_output = self._strategy.optimizer_step(self._optimizer, closure, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 239, in optimizer_step
    return self.precision_plugin.optimizer_step(optimizer, model=model, closure=closure, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/fsdp.py", line 153, in optimizer_step
    return super().optimizer_step(optimizer, model=model, closure=closure, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/precision.py", line 122, in optimizer_step
    return optimizer.step(closure=closure, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/optim/lr_scheduler.py", line 75, in wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/optim/optimizer.py", line 391, in wrapper
    out = func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/apex/optimizers/fused_adam.py", line 140, in step
    loss = closure()
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/precision.py", line 109, in _wrap_closure
    self._after_closure(model, optimizer)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/precision.py", line 88, in _after_closure
    self._clip_gradients(
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/precision.py", line 135, in _clip_gradients
    call._call_lightning_module_hook(
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 157, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
  File "/opt/NeMo/nemo/collections/nlp/models/language_modeling/megatron_base_model.py", line 630, in configure_gradient_clipping
    grad_norm = clip_grad_norm_fp32(parameters=parameters, max_norm=clip_val, use_fsdp=self.use_fsdp,)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/opt/NeMo/nemo/collections/nlp/modules/common/megatron/clip_grads.py", line 180, in clip_grad_norm_fp32
    torch._foreach_mul_(grads, clip_coeff_clamped[0])
IndexError: invalid index of a 0-dim tensor. Use `tensor.item()` in Python or `tensor.item<T>()` in C++ to convert a 0-dim tensor to a number
```

**Review by Victor49152 — CHANGES_REQUESTED** [review]

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by Victor49152** [comment]
jenkins

**Review by Victor49152 — APPROVED** [review]

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Comment by mwawrzos** [comment]
jenkins

**Pull request merged**